### ISSUE TITLE
Building

### ISSUE THREAD

**Issue by mlutze** [stephentetley/flix-sandbox#1]
I'm trying to build this package since it's a dependency of your flix-parsec package that I'd like to use for a small project. However, I'm not able to get the workflow to run properly. Do you have a build script (or just a set of steps you use) for building the Java jar and Flix package?

**Comment by mlutze** [comment]
@stephentetley (pinging since I'm not sure the ticket creates a notification)

**Comment by stephentetley** [comment]
Hi Matthew - I haven't worked out how to build / compile packages yet. For flix-sandbox which brings in third party jars, I add the source and jars to my Flix compiler project and build a whole `flix.jar` - source dependencies are added to `main/src/ca/waterloo/flix/api/Flix.scala` and library dependencies (the jars) are added to `build.gradle` and added as libraries within IntelliJ.

I'm not sure the best way to progress - Flix-parsec depends heavily on the regex code in flix-sandbox which is quite "alpha" but it could be messy to remove. The Chain module in flix-sandbox could be moved into an internal module in flix-parsec.

**Comment by mlutze** [comment]
Hi Stephen, thanks for the reply.

Just for your information, you should be able to include jars in your build (without rebuilding the compiler) by doing something like this:
```bash
java -cp "MyUtils.jar:flix.jar" ca.uwaterloo.flix.Main <flix args>
#         ^^^ colon-separated jars, including flix.jar
#            (I was just informed that it's semicolon-separated on Windows)
```

For building a "proper" Flix package, the packager is very strict (with poor error messages), but you should be able to see what is required by doing `flix init` in an empty folder, creating:
* `build` directory
* `HISTORY.md`
* `lib` directory
* `LICENSE.md`
* `README.md`
* `src` directory
* `test` directory

With those in place you can do `flix build-pkg` (or rather, the modified command as above with `build-pkg`). (Then if you create a release on GitHub with a semantic version tag (e.g. `v0.1.0`), users can consume your package with `flix install stephentetley/flix-sandbox`).

The packager is still in an early state, so all that is mostly undocumented and subject to change.

**Comment by stephentetley** [comment]
Thanks Matthew - I'll look into that.

At the moment, I'm trying to minimize flix-parsec's dependencies. I should have a PR to add Chain to the stdlib this week. Then I can split Regex out of flix-sandbox (temporarily changing functions that use Streams to use Lists) so it should only depend on the stdlib (with Chain). Then I'll look to putting the parts for flix-parsec that still depend on flix-sandbox into an optional module so the core can be used without flix-sandbox.

BTW are you looking to parse JSON? I've a parser written with flix-parsec here: 

https://github.com/stephentetley/basic-json

**Comment by mlutze** [comment]
Sounds great, thanks!

My little naive JSON parser was for a different project. I'll definitely be consuming yours if/when I come back to that one.

**Comment by stephentetley** [comment]
I've made a new repo for the regex modules from `flix-sandbox` so they can be used more easily:

https://github.com/stephentetley/flix-regex

`flix-regex` should just depend on `Chain` from `flix-sandbox`, which I've sumbitted as a PR for Flix's stdlib.

I've updated (but not really tested) `flix-parsec` so the core modules should only depend on `Chain` + `flix-regex`. The file IO code that depends on more of `flix-sandbox` has been moved to an optional module `FlixParsec/Extras`.
